src/systems/filecoin_files/data_transfer/data_transfer_subsystem.id

```diff
@@ -1,66 +1,134 @@
-    sent           UVarint
-    received       UVarint
-
-    channelType()  DataTransferType  @(cached)
-    transferNum()  Float             @(cached)
+	// an identifier for this channel shared by request and responder, set by requestor through protocol
+	transferID     TransferID
+	// base CID for the piece being transferred
+	PieceRef      CID
+	// portion of Piece to return, specified by an IPLD selector
+	Selector      ipld.Selector
+	// used to verify this channel
+	voucher       DataTransferVoucher
+	// the party that is sending the data (not who initiated the request)
+	sender        libp2p.PeerID
+	// the party that is receiving the data (not who initiated the request)
+	recipient     libp2p.PeerID
+	// expected amount of data to be transferred
+	totalSize     UVarint
 }
 
-type DataTransferParams struct {
-    channelType    DataTransferType
-    contentID      CID
-    dataIncrement  UVarint
-    sender         Address
-    recipient      Address
-    totalSize      Address
+// DataTransferState is immutable channel data plus mutable state
+type DataTransferState struct @(mutable) {
+	DataTransferChannel
+	// total bytes sent from this node (0 if receiver)
+	sent          UVarint
+	// total bytes received by this node (0 if sender)
+	received      UVarint
 }
 
-type DataTransferType union {
-    Sender
-    Recipient
+type Open struct {}
+type Progress struct {}
+type Error struct {}
+type Complete struct {}
+
+type DataTransferEvent union {
+	Open
+	Progress
+	Error
+	Complete
 }
 
-type Sender struct {}  // TODO: remove this
-type Recipient struct {}  // TODO: remove this
+type DataTransferSubscriber struct {
+	OnEvent(event DataTransferEvent, channelState DataTransferState) struct {}
+}
 
-type Ongoing struct {}
-type Completed struct {}
-type Failed struct {}
-type ChannelNotFoundError struct {}
-type DataTransferStatus union {
-    Ongoing
-    Completed
-    Failed
-    ChannelNotFoundError
+// RequestValidator is an interface implemented by the client of the data transfer module to validate requests
+type RequestValidator struct {
+	ValidatePush(
+		sender libp2p.PeerID, 
+		voucher DataTransferVoucher, 
+		PieceRef CID, 
+		Selector ipld.Selector) struct {}
+	ValidatePull(		
+		receiver libp2p.PeerID, 
+		voucher DataTransferVoucher, 
+		PieceRef CID, 
+		Selector ipld.Selector) struct {}
+}
+
+type DataTransferSubsystem struct @(mutable) {
+	host					libp2p.Node
+	dataTransfers {ChannelID: DataTransferState}
+	requestValidator RequestValidator
+	pieceStore    piece.PieceStore
+
+	// open a data transfer that will send data to the recipient peer and
+	// transfer the whole piece
+	OpenPushDataChannel(to libp2p.PeerID, voucher DataTransferVoucher, PieceRef CID) ChannelID
+
+	// open a data transfer that will send data to the recipient peer and
+	// transfer parts of the piece that match the selector
+	OpenPartialPushDataChannel(to libp2p.PeerID, voucher DataTransferVoucher, PieceRef CID, Selector ipld.Selector) ChannelID
```

Comment:
The distinction between these two is working around a limitation of IPLD selectors, which IMO are too opinionated about how they can be used. However, even if an unbounded recursive selector is not made available, can we not always know a good upper bound anyway? When sending data to a miner, the client knows the complete data depths and provides the selector to the miner.

Reply:
yes, I am inclined to simplify these methods and will update